### PR TITLE
8279651: [vectorapi] Implement the missing intrinsics for casting from integrals on x64

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch implements the remaining casts from integral types on x64.

Thank you very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8279651](https://bugs.openjdk.java.net/browse/JDK-8279651): [vectorapi] Implement the missing intrinsics for casting from integrals on x64

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 2b5c1e9b847ea53832cfddac41b174596ead158d

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7002/head:pull/7002` \
`$ git checkout pull/7002`

Update a local copy of the PR: \
`$ git checkout pull/7002` \
`$ git pull https://git.openjdk.java.net/jdk pull/7002/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7002`

View PR using the GUI difftool: \
`$ git pr show -t 7002`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7002.diff">https://git.openjdk.java.net/jdk/pull/7002.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
This looks good. I will test it.

**Comment by vnkozlov** [comment]
I got failure in next tests when run with `-XX:UseAVX=3 -XX:+UnlockDiagnosticVMOptions -XX:+UseKNLSetting`:

compiler/vectorapi/reshape/TestVectorCastAVX512.java
compiler/codegen/TestLongDoubleVect.java
jdk/incubator/vector/VectorMaxConversionTests.java
jdk/incubator/vector/Vector64ConversionTests.java#id0
jdk/incubator/vector/Vector128ConversionTests.java
jdk/incubator/vector/Vector256ConversionTests.java
```
#  Internal Error (/workspace/open/src/hotspot/cpu/x86/assembler_x86.cpp:2703), pid=31504, tid=31619
#  Error: assert(VM_Version::supports_avx512dq()) failed

Current CompileTask:
C2:   1647  456             jdk.incubator.vector.AbstractVector$$Lambda$147/0x0000000801087788::apply (15 bytes)

Stack: [0x00007f8fc59a7000,0x00007f8fc5aa8000],  sp=0x00007f8fc5aa2720,  free space=1005k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x66a494]  Assembler::kortestbl(KRegisterImpl*, KRegisterImpl*)+0x134
V  [libjvm.so+0x8b3ba1]  C2_MacroAssembler::vector_castL2FD(XMMRegisterImpl*, XMMRegisterImpl*, XMMRegisterImpl*, XMMRegisterImpl*, RegisterImpl*, KRegisterImpl*, BasicType, int, int)+0x141
V  [libjvm.so+0x4316a0]  vcastLtoFD_evexNode::emit(CodeBuffer&, PhaseRegAlloc*) const+0x280
V  [libjvm.so+0x16484f8]  PhaseOutput::scratch_emit_size(Node const*)+0x438
V  [libjvm.so+0x163e968]  PhaseOutput::shorten_branches(unsigned int*)+0x2e8
V  [libjvm.so+0x165159a]  PhaseOutput::Output()+0xc6a
V  [libjvm.so+0xa93197]  Compile::Code_Gen()+0x427
V  [libjvm.so+0xa9f50e]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x166e
```

**Comment by merykitty** [comment]
Thanks a lot for your testing and review, the assert should have been fixed now.

**Comment by vnkozlov** [comment]
New failure. Test `compiler/codegen/TestLongDoubleVect.java` produces wrong results when run with `-XX:UseAVX=3 -XX:+UnlockDiagnosticVMOptions -XX:+UseKNLSetting`
```
Testing Long + Double vectors
Warmup
Verification
----------System.err:(985/53059)----------
test_conv_l2d: a1[1] = 0.0 != -9.223372036854776E18
test_conv_l2d: a1[2] = 0.0 != -9.223372036854776E18
test_conv_l2d: a1[3] = 0.0 != -9.223372036854776E18
...
```

**Comment by merykitty** [comment]
Oh silly me, it should have been fixed now.
Thanks.

**Comment by vnkozlov** [comment]
I submitted new testing.  You need second review.

**Comment by PaulSandoz** [comment]
@jatin-bhateja or @sviswa7 can you take a look?

**Review by vnkozlov — APPROVED** [review]
Tier1-3 testing passed.

File: src/hotspot/cpu/x86/x86.ad
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86.ad:6951]: The following code is redundant:
```
  switch (to_elem_bt) {
      case T_BYTE:
        if (!VM_Version::supports_avx512vl()) {
          vlen_enc = Assembler::AVX_512bit;
        }
```

File: src/hotspot/cpu/x86/x86.ad
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86.ad:6952]: Same question here (as for `VectorCastB2X`): why doesn't it match the condition removed from `Matcher::match_rule_supported_vector()` (e.g., `T_FLOAT` vs `T_DOUBLE`)?

File: src/hotspot/cpu/x86/x86.ad
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86.ad:1787]: I don't see any new code added for `VectorCastI2X`. Why is it so?

File: src/hotspot/cpu/x86/x86.ad
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86.ad:7150]: `!VM_Version::supports_avx512dq()` is equivalent to `UseAVX <= 2 || !VM_Version::supports_avx512dq()` here, isn't it?

Comment (iwanowww) [reply src/hotspot/cpu/x86/x86.ad:7150]:
```
case Op_VectorCastL2X:
   ...
   } else if (!is_integral_type(bt) && !VM_Version::supports_avx512dq()) {
     return false;
   }
```
Overall, It becomes harder and harder to ensure the relevant AD instructions are exhaustive w.r.t. predicates and supported modes. 

I'd expect to see different variations of `is_floating_point_type(Matcher::vector_element_basic_type(n))` and`VM_Version::supports_avx512dq())` to enumerate all important configurations.

File: src/hotspot/cpu/x86/x86.ad
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86.ad:1793]: No new code for `VectorCastL2X` to integral types? Why?

**Review by iwanowww — CHANGES_REQUESTED** [review]
Overall,

**Comment by merykitty** [comment]
@iwanowww Thanks for your reviews, I think I will update the PR this weekend. In the meantime, I would reply to your concerns.

The condition for a vector node to be generated is that

    input shape supported  && // this is implicit, happen for example at the intrinsic of VectorSupport.convert, I think we should make it more explicit
    output shape supported && // check at the start of Matcher::match_rule_supported_vector
    special rules             // List explicitly in the method

Detailed logic for each circumstance will be responded separately.

Thank you very much.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4150]: Good idea, do you know how to properly save the state of a `xmm` register? Currently I can only see using x86_32 approach which does not need to touch the `xmm` registers.

File: src/hotspot/cpu/x86/x86.ad
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:6881]: Because on AVX1, 256-bit integral vectors are not supported, so the only candidates for a 256-bit vector here are float and double.

File: src/hotspot/cpu/x86/x86.ad
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:6951]: Yes it is. However, the intention of this code seems that `vpmovwb` is promoted to `AVX_512bit` when avx512vl is not available, and the predicate is too restrictive. I think this needs a closer look.

File: src/hotspot/cpu/x86/x86.ad
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:1787]: This part is redundant, as above the only candidates for 256-bit vector on AVX1 is float and double. However, to cast to a 256-bit float vector we need a 256-bit int vector which is not supported, so a <= 128-bit int vector can only be casted to a 256-bit vector of type double.

File: src/hotspot/cpu/x86/x86.ad
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:1793]: This is also redundant, the only long vector that can appear in AVX1 is <= 128-bit, so we can't cast to 256-bit integral vectors here.

File: src/hotspot/cpu/x86/x86.ad
Comment (iwanowww) [reply src/hotspot/cpu/x86/x86.ad:6951]: AVX512VL support always comes along with AVX512BW in Intel CPUs. Xeon Phis didn't support VL, but they didn't support BW as well. So, it's safe to assume that BW implies VL and use `VM_Version::supports_avx512vlbw()` instead of `VM_Version::supports_avx512bw()`. It makes the aforementioned code effectively dead.

**Review by iwanowww — COMMENTED** [review]
Thanks for the clarifications.

**Pull request closed without merging**
